clippy_utils/src/eager_or_lazy.rs

```diff
@@ -9,128 +9,191 @@
+                    Res::Def(_, id) => match path {
+                        QPath::Resolved(_, p) => {
+                            self.eagerness |= fn_eagerness(self.cx, id, p.segments.last().unwrap().ident.name, args);
+                        },
+                        QPath::TypeRelative(_, name) => {
+                            self.eagerness |= fn_eagerness(self.cx, id, name.ident.name, args);
+                        },
+                        QPath::LangItem(..) => self.eagerness = Lazy,
+                    },
+                    _ => self.eagerness = Lazy,
+                },
+                // No need to walk the arguments here
+                ExprKind::MethodCall(..) if is_const_expr(self.cx, e) => return,
+                ExprKind::MethodCall(name, _, args, _) => {
+                    self.eagerness |= self
+                        .cx
+                        .typeck_results()
+                        .type_dependent_def_id(e.hir_id)
+                        .map_or(Lazy, |id| fn_eagerness(self.cx, id, name.ident.name, args));
+                },
+                ExprKind::Index(_, e) => {
+                    let ty = self.cx.typeck_results().expr_ty_adjusted(e);
+                    if is_copy(self.cx, ty) && !ty.is_ref() {
+                        self.eagerness |= NoChange;
+                    } else {
+                        self.eagerness = Lazy;
+                    }
+                },
+                ExprKind::Unary(UnOp::Deref, e) if !self.cx.typeck_results().expr_ty(e).is_unsafe_ptr() => (),
+                ExprKind::Unary(UnOp::Deref, _) => self.eagerness |= NoChange,
+                ExprKind::Unary(_, e)
+                    if matches!(
+                        self.cx.typeck_results().expr_ty(e).kind(),
+                        ty::Bool | ty::Int(_) | ty::Uint(_),
+                    ) => {},
+                ExprKind::Binary(_, lhs, rhs)
+                    if matches!(
+                        (
+                            self.cx.typeck_results().expr_ty(lhs).kind(),
+                            self.cx.typeck_results().expr_ty(rhs).kind(),
+                        ),
+                        (ty::Bool | ty::Int(_) | ty::Uint(_), ty::Bool | ty::Int(_) | ty::Uint(_)),
+                    ) => {},
+
+                ExprKind::Break(..)
+                | ExprKind::Continue(_)
+                | ExprKind::Ret(_)
+                | ExprKind::InlineAsm(_)
+                | ExprKind::LlvmInlineAsm(_)
+                | ExprKind::Yield(..)
+                | ExprKind::Err => {
+                    self.eagerness = ForceNoChange;
+                    return;
+                },
+                ExprKind::Box(_)
+                | ExprKind::Unary(..)
+                | ExprKind::Binary(..)
+                | ExprKind::Loop(..)
+                | ExprKind::Block(Block { stmts: [_, ..], .. }, _) => self.eagerness = Lazy,
+                _ => (),
```

Comment:
At that point it may as well be an exhaustive match. `ForceNoChange` is for when the meaning of the code would be changed.

Reply:
It's a matter of defensive programming. If there's a pattern we forgot, or if a variant of ExprKind is added, the default behavior should be to not lint.

Reply:
True. Having an exhaustive match would cause the same thing and make sure someone actually makes a decision when a new variant is added.

In case it wasn't clear I was suggesting this instead, not saying it wouldn't be a better default.

Reply:
Thanks for clarifying. I'd be fine with either approach.